File: Data Structures & Algorithms/queue/submission-2.cpp
class Node{
public:
    int val;
    Node* next;
    Node* prev;
    Node(int val): val(val), next(nullptr), prev(nullptr){}
    Node(int v, Node* p, Node* n): val(v), next(n), prev(p){}
};

class Deque {
private:
    Node* head;
    Node* tail;
    int size;
public:
    Deque() {
        head = new Node(-1);
        tail = new Node(-1);
        head->next = tail;
        tail->prev = head;    
        size = 0;
    }

    bool isEmpty() {
        return head->next == tail;
    }

    void append(int value) {
        Node* node = new Node(value, tail->prev, tail);
        tail->prev->next = node;
        tail->prev = node;
        size++;
    }

    void appendleft(int value) {
        Node* node = new Node(value, head, head->next);
        head->next->prev = node;
        head->next = node;
        size++;
    }

    int pop() {
        if (size == 0){
            return -1;
        }
        Node* node = tail->prev;
        tail->prev = node->prev;
        node->prev->next = tail;
        size--;
        return node->val;
    }

    int popleft() {
        if (size == 0){
            return -1;
        }
        Node* node = head->next;
        head->next = node->next;
        node->next->prev = head;
        size--;
        return node->val;
    }
};

File: Data Structures & Algorithms/singlyLinkedList/submission-2.py
class Node:

    def __init__(self, val = 0):
        self.val = val
        self.next = None

class LinkedList:
    
    def __init__(self):
        self.head = Node()
        self.tail = self.head
    
    def get(self, index: int) -> int:
        curr = self.head.next
        while curr and index > 0:
            curr = curr.next
            index -= 1

        if not curr:
            return -1

        return curr.val

    def insertHead(self, val: int) -> None:
        node = Node(val)
        node.next = self.head.next
        self.head.next = node
        if self.tail == self.head:
            self.tail = node

    def insertTail(self, val: int) -> None:
        self.tail.next = Node(val)
        self.tail = self.tail.next     


    def remove(self, index: int) -> bool:
        prev = self.head
        while prev and prev.next and index > 0:
            prev = prev.next
            index -= 1

        if not prev or not prev.next:
            return False

        if prev.next == self.tail:
            self.tail = prev
            
        prev.next = prev.next.next
        return True

    def getValues(self) -> List[int]:
        arr = []
        curr = self.head.next
        while curr:
            arr.append(curr.val)
            curr = curr.next

        return arr
        
